plugins/hls-signature-help-plugin/src/Ide/Plugin/SignatureHelp.hs

```diff
@@ -0,0 +1,208 @@
+                                                       smallestContainingSatisfying,
+                                                       sourceNodeInfo)
+import           Development.IDE.GHC.Compat.Util      (LexicalFastString (LexicalFastString))
+import           GHC.Data.Maybe                       (rightToMaybe)
+import           GHC.Types.SrcLoc                     (isRealSubspanOf)
+import           Ide.Plugin.Error                     (getNormalizedFilePathE)
+import           Ide.Types                            (PluginDescriptor (pluginHandlers),
+                                                       PluginId,
+                                                       PluginMethodHandler,
+                                                       defaultPluginDescriptor,
+                                                       mkPluginHandler)
+import           Language.LSP.Protocol.Message        (Method (Method_TextDocumentSignatureHelp),
+                                                       SMethod (SMethod_TextDocumentSignatureHelp))
+import           Language.LSP.Protocol.Types          (Null (Null),
+                                                       ParameterInformation (ParameterInformation),
+                                                       Position (Position),
+                                                       SignatureHelp (SignatureHelp),
+                                                       SignatureHelpParams (SignatureHelpParams),
+                                                       SignatureInformation (SignatureInformation),
+                                                       TextDocumentIdentifier (TextDocumentIdentifier),
+                                                       UInt,
+                                                       type (|?) (InL, InR))
+
+data Log = LogDummy
+
+instance Pretty Log where
+    pretty = \case
+        LogDummy -> "TODO(@linj) remove this dummy log"
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor _recorder pluginId =
+    (defaultPluginDescriptor pluginId "Provides signature help of something callable")
+        { Ide.Types.pluginHandlers = mkPluginHandler SMethod_TextDocumentSignatureHelp signatureHelpProvider
+        }
+
+-- TODO(@linj) get doc
+signatureHelpProvider :: PluginMethodHandler IdeState Method_TextDocumentSignatureHelp
+signatureHelpProvider ideState _pluginId (SignatureHelpParams (TextDocumentIdentifier uri) position _mProgreeToken _mContext) = do
+    nfp <- getNormalizedFilePathE uri
+    mResult <- runIdeActionE "signatureHelp" (shakeExtras ideState) $ do
+        -- TODO(@linj) why HAR {hieAst} may have more than one AST?
+        (HAR {hieAst, hieKind}, positionMapping) <- useWithStaleFastE GetHieAst nfp
+        case fromCurrentPosition positionMapping position of
+            Nothing -> pure Nothing
+            Just oldPosition -> do
+                let functionName =
```

Comment:
you probably want to do all these in one go, this looks like it will traverse the AST three times! I think you can just do one call to `extractInfoFromSmallestContainingFunctionApplicationAst` and get all the pieces together though?